File: packages/cordova-plugin-actito-in-app-messaging/src/android/ActitoInAppMessagingPlugin.kt
package com.actito.iam.cordova

import org.apache.cordova.CallbackContext
import org.apache.cordova.CordovaArgs
import org.apache.cordova.CordovaPlugin
import org.apache.cordova.PluginResult
import org.json.JSONArray
import org.json.JSONObject
import com.actito.Actito
import com.actito.iam.ActitoInAppMessaging
import com.actito.iam.ktx.inAppMessaging
import com.actito.iam.models.ActitoInAppMessage

class ActitoInAppMessagingPlugin : CordovaPlugin(), ActitoInAppMessaging.MessageLifecycleListener {

    override fun pluginInitialize() {
        logger.hasDebugLoggingEnabled = Actito.options?.debugLoggingEnabled ?: false

        Actito.inAppMessaging().addLifecycleListener(this)
    }

    override fun onDestroy() {
        Actito.inAppMessaging().removeLifecycleListener(this)
    }

    override fun execute(action: String, args: CordovaArgs, callback: CallbackContext): Boolean {
        when (action) {
            "hasMessagesSuppressed" -> hasMessagesSuppressed(args, callback)
            "setMessagesSuppressed" -> setMessagesSuppressed(args, callback)

            // Event broker
            "registerListener" -> registerListener(args, callback)

            else -> {
                callback.error("No implementation for action '$action'.")
                return false
            }
        }

        return true
    }

    // region Actito In-App Messaging

    private fun hasMessagesSuppressed(@Suppress("UNUSED_PARAMETER") args: CordovaArgs, callback: CallbackContext) {
        callback.success(Actito.inAppMessaging().hasMessagesSuppressed)
    }

    private fun setMessagesSuppressed(@Suppress("UNUSED_PARAMETER") args: CordovaArgs, callback: CallbackContext) {
        try {
            val suppressed = args.getBoolean(0)
            val evaluateContext =
                if (!args.isNull(1)) {
                    args.getBoolean(1)
                } else {
                    false
                }

            Actito.inAppMessaging().setMessagesSuppressed(suppressed, evaluateContext)

            callback.void()
        } catch (e: Exception) {
            callback.error(e.message)
        }
    }

    // endregion

    // region ActitoInAppMessaging.MessageLifecycleListener

    override fun onMessagePresented(message: ActitoInAppMessage) {
        try {
            ActitoInAppMessagingPluginEventBroker.dispatchEvent("message_presented", message.toJson())
        } catch (e: Exception) {
            logger.error("Failed to emit the message_presented event.", e)
        }
    }

    override fun onMessageFinishedPresenting(message: ActitoInAppMessage) {
        try {
            ActitoInAppMessagingPluginEventBroker.dispatchEvent("message_finished_presenting", message.toJson())
        } catch (e: Exception) {
            logger.error("Failed to emit the message_finished_presenting event.", e)
        }
    }

    override fun onMessageFailedToPresent(message: ActitoInAppMessage) {
        try {
            ActitoInAppMessagingPluginEventBroker.dispatchEvent("message_failed_to_present", message.toJson())
        } catch (e: Exception) {
            logger.error("Failed to emit the message_failed_to_present event.", e)
        }
    }

    override fun onActionExecuted(message: ActitoInAppMessage, action: ActitoInAppMessage.Action) {
        try {
            val json = JSONObject()
            json.put("message", message.toJson())
            json.put("action", action.toJson())

            ActitoInAppMessagingPluginEventBroker.dispatchEvent("action_executed", json)
        } catch (e: Exception) {
            logger.error("Failed to emit the action_executed event.", e)
        }
    }

    override fun onActionFailedToExecute(
        message: ActitoInAppMessage,
        action: ActitoInAppMessage.Action,
        error: Exception?
    ) {
        try {
            val json = JSONObject()
            json.put("message", message.toJson())
            json.put("action", action.toJson())

            if (error != null) {
                json.put("error", error.message)
            }

            ActitoInAppMessagingPluginEventBroker.dispatchEvent("action_failed_to_execute", json)
        } catch (e: Exception) {
            logger.error("Failed to emit the action_failed_to_execute event.", e)
        }
    }

    // endregion

    private fun registerListener(@Suppress("UNUSED_PARAMETER") args: CordovaArgs, callback: CallbackContext) {
        ActitoInAppMessagingPluginEventBroker.setup(preferences, object : ActitoInAppMessagingPluginEventBroker.Consumer {
            override fun onEvent(event: ActitoInAppMessagingPluginEventBroker.Event) {
                val payload = JSONObject()
                payload.put("name", event.name)
                when (event.payload) {
                    null -> {} // Skip encoding null payloads.
                    is Boolean -> payload.put("data", event.payload)
                    is Int -> payload.put("data", event.payload)
                    is Float -> payload.put("data", event.payload)
                    is Double -> payload.put("data", event.payload)
                    is String -> payload.put("data", event.payload)
                    is JSONObject -> payload.put("data", event.payload)
                    is JSONArray -> payload.put("data", event.payload)
                    else -> throw IllegalArgumentException("Unsupported event payload of type '${event.payload::class.java.simpleName}'.")
                }

                val result = PluginResult(PluginResult.Status.OK, payload)
                result.keepCallback = true

                callback.sendPluginResult(result)
            }
        })
    }
}

private fun CallbackContext.void() {
    sendPluginResult(PluginResult(PluginResult.Status.OK, null as String?))
}

private fun CallbackContext.success(b: Boolean) {
    sendPluginResult(PluginResult(PluginResult.Status.OK, b))
}

File: packages/cordova-plugin-actito-in-app-messaging/src/ios/ActitoInAppMessagingPlugin.swift
import ActitoKit
import ActitoInAppMessagingKit

@MainActor
@objc(ActitoInAppMessagingPlugin)
class ActitoInAppMessagingPlugin : CDVPlugin {

    override func pluginInitialize() {
        super.pluginInitialize()

        Actito.shared.inAppMessaging().delegate = self
    }

    @objc func registerListener(_ command: CDVInvokedUrlCommand) {
        ActitoInAppMessagingPluginEventBroker.startListening(settings: commandDelegate.settings as? [AnyHashable: Any]) { event in
            var payload: [String: Any] = [
                "name": event.name,
            ]

            if let data = event.payload {
                payload["data"] = data
            }

            let result = CDVPluginResult(status: .ok, messageAs: payload)
            result.keepCallback = true

            self.commandDelegate!.send(result, callbackId: command.callbackId)
        }
    }

    // MARK: - Actito In-App Messaging

    @objc func hasMessagesSuppressed(_ command: CDVInvokedUrlCommand) {
        let result = CDVPluginResult(status: .ok, messageAs: Actito.shared.inAppMessaging().hasMessagesSuppressed)
        self.commandDelegate!.send(result, callbackId: command.callbackId)
    }

    @objc func setMessagesSuppressed(_ command: CDVInvokedUrlCommand) {
        let suppressed = command.argument(at: 0) as! Bool
        let evaluateContext = command.argument(at: 1) as? Bool ?? false

        Actito.shared.inAppMessaging().setMessagesSuppressed(suppressed, evaluateContext: evaluateContext)

        let result = CDVPluginResult(status: .ok)
        self.commandDelegate!.send(result, callbackId: command.callbackId)
    }
}

extension ActitoInAppMessagingPlugin: ActitoInAppMessagingDelegate {
    func actito(_ actito: ActitoInAppMessaging, didPresentMessage message: ActitoInAppMessage) {
        do {
            ActitoInAppMessagingPluginEventBroker.dispatchEvent(
                name: "message_presented",
                payload: try message.toJson()
            )
        } catch {
            logger.error("Failed to emit the message_presented event.", error: error)
        }
    }

    func actito(_ actito: ActitoInAppMessaging, didFinishPresentingMessage message: ActitoInAppMessage) {
        do {
            ActitoInAppMessagingPluginEventBroker.dispatchEvent(
                name: "message_finished_presenting",
                payload: try message.toJson()
            )
        } catch {
            logger.error("Failed to emit the message_finished_presenting event.", error: error)
        }
    }

    func actito(_ actito: ActitoInAppMessaging, didFailToPresentMessage message: ActitoInAppMessage) {
        do {
            ActitoInAppMessagingPluginEventBroker.dispatchEvent(
                name: "message_failed_to_present",
                payload: try message.toJson()
            )
        } catch {
            logger.error("Failed to emit the message_failed_to_present event.", error: error)
        }
    }

    func actito(_ actito: ActitoInAppMessaging, didExecuteAction action: ActitoInAppMessage.Action, for message: ActitoInAppMessage) {
        do {
            ActitoInAppMessagingPluginEventBroker.dispatchEvent(
                name: "action_executed",
                payload: [
                    "message": try message.toJson(),
                    "action": try action.toJson(),
                ]
            )
        } catch {
            logger.error("Failed to emit the action_executed event.", error: error)
        }
    }

    func actito(_ actito: ActitoInAppMessaging, didFailToExecuteAction action: ActitoInAppMessage.Action, for message: ActitoInAppMessage, error: Error?) {
        do {
            var payload: [String: Any] = [
                "message": try message.toJson(),
                "action": try action.toJson(),
            ]

            if let error = error {
                payload["error"] = error.localizedDescription
            }

            ActitoInAppMessagingPluginEventBroker.dispatchEvent(
                name: "action_failed_to_execute",
                payload: payload
            )
        } catch {
            logger.error("Failed to emit the action_failed_to_execute event.", error: error)
        }
    }
}
